lib/src/package_analysis/summary.dart

```diff
@@ -0,0 +1,49 @@
+import 'dart:io';
+
+import 'package:analyzer/dart/analysis/context_builder.dart';
+import 'package:analyzer/dart/analysis/context_locator.dart';
+import 'package:analyzer/dart/analysis/results.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:path/path.dart' as path;
+
+import './common.dart';
+
+Future<void> generateSummary(String packageLocation) async {
+  if (!await FileSystemEntity.isDirectory(packageLocation)) {
+    exitFatal('Please specify a directory for analysis.', 1);
+  }
+
+  // initialise analyser
+  final context = ContextBuilder().createContext(
+    contextRoot: ContextLocator().locateRoots(
+      // use canonicalize to ensure we get an absolute path
+      includedPaths: [path.canonicalize(packageLocation)],
+    ).first,
+  );
+
+  for (final filePath in context.contextRoot.analyzedFiles()) {
+    final session = context.currentSession;
+    final someResolvedLibrary = await session.getResolvedLibrary(filePath);
+
+    // match [packageLocation]/lib/*.dart
+    // but exclude [packageLocation]/lib/src/*.dart
+    if (!(path.isWithin(path.join(packageLocation, 'lib'), filePath) &&
+        !path.isWithin(path.join(packageLocation, 'lib', 'src'), filePath) &&
+        path.extension(filePath) == '.dart')) {
+      continue;
+    }
+
+    // ensure that resolving has been successful
+    if (someResolvedLibrary is! ResolvedLibraryResult) {
+      stderr.writeln('$filePath is not a library');
```

Comment:
> Isn't it more like that analysis of the specified file failed?

More specifically, I might say 'Building a resolved AST for this library returned an invalid result.' (Not sure if this is too verbose)



> Then people calling `generateSummary` can provide an instance of `PackageAnalysisContext` and they can decide whether to print to stderr or write a log file.

Right, so they would have to provide their own handler? In case of a fatal error I could throw an exception instead of calling the handler (well, I suppose I could call the handler, let the user log the error and THEN throw the exception), like `InconsistentAnalysisException` in the case of `AnalysisSession`.